### PR TITLE
Frontend

### PR DESCRIPTION
**Pull request by heruka-urgyen**
### Scope
- [x] add front-end dependencies
- [x] add front-end dev mode
- [x] render articles
- [x] resolve questions regarding data presentation (see below)
- [x] add login flow
- [x] fetch and store labels
- [x] refactor
- [ ] add build mode integrated with backend (https://github.com/bzz/scholar-alert-digest/issues/20)
- [x] update readme

### Run in dev mode
1. `cd frontend`
2. `npm i`
3. `npm run dev`
4. open `localhost:9000` in browser

### Questions
1. Get data for unread emails? 
2. What is `read` / `unread` in data and how to move articles between them?
3. What are unique paper titles?
4. How should refs be rendered? It's returned as a list, but all examples have only one. What is the number inside parentheses? Again, in all examples it's always 1.
5. What is `Freq` in data?
6. How should we approach bundling everything together? I can build to a dir with all assets and scripts. Perhaps we need a higher level automation that will trigger front-end build.
7. Shall we update readme with info about front-end, e.g. how to run it in dev / prod mode? Or shall I make a separate readme under frontend dir?

### Preview
poc

### CODE REVIEW

File: frontend/src/index.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+          {data.unread.length + data.read.length}
+        </li>
+        <li>
+          <b>Unique paper titles: </b>
+          <Alert />
+        </li>
+      </ul>
+      <h2>New papers</h2>
+      <ul>
+        {data.unread.map(article => (
```
Comment (bzz) [inline frontend/src/index.js:34]: For consistency, may be we should prefer a "paper" instead or "article", same as on the backend.
May be we could aslo introduce a single paper component (that can be in collapsed and uncollapsed states) right from the beginning, even if it is in the same file?

**Review by bzz — COMMENTED** [review]
I will try to answer all questions about data-model (1-5) in a simple doc a bit later.

On 6: IMO ATM in this or next PR it would be enough to have a frontend build profile in npm that outputs all nessesary resources to a dir. 
I.e. it's totally fine to expect having everyone to run it locally in the dev mode for now.
As soon as we remove all impediments for having an actual deployment (#20 ) - I will add a `build.sh` that triggered it and a backend build, bundles all the resources into the single binary and add a route serving them.

On 7: We definitely should have a readme inside the frontend dir, with the dev mode instructions.

We should also add a section to the root readme, pointing to the frontend dir and mentioning the status (experiential) of our new, awesome, interactive GUI.

File: frontend/src/index.ejs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8">
+    <link rel="icon" href="http://emojipedia-us.s3.dualstack.us-west-1.amazonaws.com/thumbs/240/apple/232/page-with-curl_1f4c3.png">
```
[the remark this answers is not in the document]
Comment (heruka-urgyen) [reply frontend/src/index.ejs:5]: Sure, it's not hard

File: frontend/src/index.js (hunk 1 above)
Comment (heruka-urgyen) [reply frontend/src/index.js:34]: Yes, I plan to refactor it into a component either way. And let's call it paper, sure.

**Comment by heruka-urgyen** [comment]
> I will try to answer all questions about data-model (1-5) in a simple doc a bit later.

I'm waiting on this to finalize the page. 

Also, let me know if you want any changes in front-end before that. Shall we consider simple UI enhancements? Maybe we could to make a short roadmap?

**Comment by bzz** [comment]
> Shall we consider simple UI enhancements?

yes!

> make a short roadmap

yes, that would be awesome! So far #19 has several thoughts on that, as well as #32 proposes improvements with regards to the source of citations.


>> I will try to answer all questions about data-model (1-5) in a simple doc a bit later.
>
> I'm waiting on this to finalize the page.

got it, will do that today ASAP

**Comment by bzz** [comment]
>>> I will try to answer all questions about data-model (1-5) in a simple doc a bit later.
>>
>> I'm waiting on this to finalize the page.
>
> got it, will do that today ASAP.

Done at #39

**Comment by bzz** [comment]
New fixtures forth, read and unread messages were added in #40 and merged to master.

**Comment by heruka-urgyen** [comment]
> New fixtures forth, read and unread messages were added in #40 and merged to master.

On master, I'm getting 
```
> go run main.go -test
flag provided but not defined: -test
```

Am I missing something?

**Comment by bzz** [comment]
> > New fixtures forth, read and unread messages were added in #40 and merged to master.
> 
> On master, I'm getting
> 
> ```
> > go run main.go -test
> flag provided but not defined: -test
> ```
> 
> Am I missing something?

No, you are not. As mentioned elsewhere, CLI does not yet support the `-test`, only server does.
At least until [this todo](https://github.com/bzz/scholar-alert-digest/blob/f6a49f7ac6f2a8a309aaf2550874463a826e321d/cmd/server/server.go#L129) is addressed (which is also a message fetching part of the https://github.com/bzz/scholar-alert-digest/issues/20). 

The only way to test CLI is to [configure](https://github.com/bzz/scholar-alert-digest#configure) and [subscribe](https://github.com/bzz/scholar-alert-digest/blob/bfd8d986d1965310e7c1c3d62f215410d69d0205/README.md#how-to-use) for some alerts from scholar.

Hopefully going to have

**Comment by heruka-urgyen** [comment]
Got it. Got confused by https://github.com/bzz/scholar-alert-digest/pull/40/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R30

**Comment by bzz** [comment]
I tried this branch on the real data and got a blank page with that in the console

Screenshot 2020-10-23 at 23 39 35

and server logs

```
go run ./cmd/server
2020/10/23 23:37:18 starting the web server at http://localhost:8080
2020/10/23 23:37:25 GET - /?json
2020/10/23 23:37:25 Redirecting to /login as three is no session
```

the reason is that GET request did not have access token and was answered with the

```
HTTP/1.1 302 Found
Content-Type: text/html; charset=utf-8
Location: /login
Date: Fri, 23 Oct 2020 21:38:35 GMT
Content-Length: 29
```

**Comment by heruka-urgyen** [comment]
The error in console is due to cors not being enabled for normal mode. Ideally, cors only makes sense in dev mode, because normal mode should bundle front-end code, eliminating the need for cors. However, if you see a use case for dev mode with real data, then sure, we need cors.

That said, we still need to redirect to login screen. I will trigger it automatically when session cookie is missing.

**Comment by bzz** [comment]
> The error in console is due to cors not being enabled for normal mode.

Right now, CORS is always enabled, but only for the JSON responses. The error is due to the 304 response not having CORS.

After #43 \w CORS enabled for all the routs the error is different now
Screenshot 2020-10-24 at 11 51 59


> That said, we still need to redirect to login screen. I will trigger it automatically when session cookie is missing.

sounds good 👍  feel free to merge #43 as you see fit

**Comment by heruka-urgyen** [comment]
I haven't figured out how to make this work. When I pass "Access-Control-Allow-Origin" header, I'm getting

```
Cross-Origin Request Blocked: 
The Same Origin Policy disallows reading the remote resource at http://localhost:8080/?json. 
(Reason: CORS request external redirect not allowed)
```

([Some information about it](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSExternalRedirectNotAllowed?utm_source=devtools&utm_medium=firefox-cors-errors&utm_campaign=default))

What we could do, is disable backend redirect and instead make an authorization routine. I think I will need to query the server for client_id, and then trigger redirect.

**Comment by bzz** [comment]
I see.

> make an authorization routine

Isn't that what `/login` is for, in a way?

I guess, Instead of re-direct (302) to `/login` I can send back 200 with a JSON containing the external URL user needs to be sent to 🤔 

Most probably the simplest way instead of refactoring existing logic on the backend - start creating new endpoints for that under `/json/*`

> query the server for client_id

I'm not 100% sure what do you mean by that as the server is stateless.

**Comment by bzz** [comment]
#44 introduces proper `/json/*` endpoints that can be used for the frontend auth flow.

Now any request without a token will result in
```
HTTP/1.1 401 Unauthorized
Content-Type: application/json
Date: Mon, 26 Oct 2020 17:39:55 GMT
Content-Length: 354

{"error":{"status":"Unauthorized","Redirect":"https://accounts.google.com/o/oauth2/auth?...."}}
```

**Comment by heruka-urgyen** [comment]
Seems like rebase went a little wrong :/

**Comment by heruka-urgyen** [comment]
We have a problem with oauth. Currently, the `redirect_uri` is `localhost:8080`, which should be fine in production mode, when code is bundled. In development mode however,we have to then manually go to `localhost:9000`, which isn't a big deal, but still differs from expected behavior. Perhaps, we could trigger another redirect from server (to `:9000`) in development mode?

**Comment by bzz** [comment]
> Perhaps, we could trigger another redirect from server (to :9000) in development mode?

From where to where?

We could add `-dev` and ask G re-direct to our `:9090` instead, if that would help

**Comment by bzz** [comment]
> Seems like rebase went a little wrong :/

yes, it seems like when re-writing a history - a number of commits from master were re-written as well.

You need to remove those from this branch and rebase it on the latest master again.

**Comment by heruka-urgyen** [comment]
> Add a server-side re-direct to the backed from what route to what route?
> 
> We could add `-dev` and ask google to re-direct the user to `:9090` instead, if that would help thinking

Here's what's happening step by step.
1. Client sends get / post to server
2. Server returns 401 unauthorized
3. Client redirects to google
4. Google redirects to `localhost:8080`
5. After that, manually navigating to `localhost:9000` shows 401 and triggers same loop again

auth

You can try it, I just committed this

**Comment by bzz** [comment]
With https://github.com/bzz/scholar-alert-digest/pull/46

```
go run ./cmd/server -cors -dev
```

redirects to :9000 after successful auth.

**Comment by bzz** [comment]
Works like a charm, after #32 was merged.
Is there anything left here, before merging?

May be adding `/static/` prefix to work with the https://github.com/bzz/scholar-alert-digest/pull/47/files#r518880714

**Comment by heruka-urgyen** [comment]
> May be adding `/static/` prefix to work with the #47

Should work as `npm run build --env baseUrl=...`

**Pull request merged**